File: multifiledownloader/src/main/java/com/github/why168/multifiledownloader/AsyncConnectCall.java
package com.github.why168.multifiledownloader;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Handler;
import android.os.Message;
import android.util.Log;

import com.github.why168.multifiledownloader.db.DataBaseUtil;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.URL;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * 连接的线程
 *
 * @author Edwin.Wu
 * @version 2017/1/16 14:15
 * @since JDK1.8
 */
public class AsyncConnectCall extends NickRunnable {
    private final Context context;
    private final Handler handler;
    private final ConcurrentHashMap<String, AsyncDownCall> downTaskMap;
    private final ExecutorService executorService;
    private final ConcurrentHashMap<String, AsyncConnectCall> connectionTaskMap;
    private DownLoadBean bean;
    private AtomicBoolean isRunning;

    @SuppressLint("SimpleDateFormat")
    public AsyncConnectCall(Context context, Handler handler,
                            ConcurrentHashMap<String, AsyncConnectCall> connectionTaskMap,
                            ConcurrentHashMap<String, AsyncDownCall> downTaskMap,
                            ExecutorService executorService,
                            DownLoadBean bean) {
        super("AndroidHttp %s", new SimpleDateFormat("yyyy-MM-dd-HH-mm-ss-SSS").format(Calendar.getInstance().getTime()));
        this.context = context;
        this.handler = handler;
        this.downTaskMap = downTaskMap;
        this.executorService = executorService;
        this.bean = bean;
        this.isRunning = new AtomicBoolean(true);
        this.connectionTaskMap = connectionTaskMap;
        this.connectionTaskMap.put(bean.id, this);
    }


    @Override
    protected void execute() {
        bean.downloadState = DownLoadState.STATE_CONNECTION.getIndex();
        DataBaseUtil.UpdateDownLoadById(context, bean);
        notifyDownloadStateChanged(bean, DownLoadState.STATE_CONNECTION.getIndex());

        HttpURLConnection connection = null;
        try {
            connection = (HttpURLConnection) new URL(bean.url).openConnection();
            connection.setRequestMethod("GET");
            connection.setConnectTimeout(Constants.CONNECT_TIME);
            connection.setReadTimeout(Constants.READ_TIME);
            int responseCode = connection.getResponseCode();
            int contentLength = connection.getContentLength();
            if (responseCode == HttpURLConnection.HTTP_OK) {
                String ranges = connection.getHeaderField("Accept-Ranges");
                if ("bytes".equalsIgnoreCase(ranges)) {
                    bean.isSupportRange = true;
                }
                bean.totalSize = contentLength;
            } else {
                bean.downloadState = DownLoadState.STATE_ERROR.getIndex();
            }

//            DataBaseUtil.UpdateDownLoadById(context, bean);
//            notifyDownloadStateChanged(bean, DownLoadState.STATE_CONNECTION);
            Log.d("Edwin", "连接成功--isSupportRange = " + bean.isSupportRange);

            // 开始下载咯
            AsyncDownCall downLoadTask = new AsyncDownCall(context, handler, bean);
            if (downTaskMap.get(bean.id) == null) {
                downTaskMap.put(bean.id, downLoadTask);
                executorService.execute(downLoadTask);

                // 移除
                connectionTaskMap.remove(bean.id);
            }
        } catch (IOException e) {
            e.printStackTrace();
            isRunning.set(false);
            bean.downloadState = DownLoadState.STATE_ERROR.getIndex();
            DataBaseUtil.UpdateDownLoadById(context, bean);
            notifyDownloadStateChanged(bean, DownLoadState.STATE_ERROR.getIndex());
            Log.d("Edwin", "连接失败");
        } finally {
            if (connection != null) {
                connection.disconnect();
            }
        }
    }

    public boolean isCanceled() {
        return isRunning.get();
    }

    public void cancel() {
        isRunning.set(true);
    }

    /**
     * 当下载状态发送改变的时候调用
     */
    private void notifyDownloadStateChanged(DownLoadBean bean, int state) {
        Message message = handler.obtainMessage();
        message.obj = bean;
        message.what = state;
        handler.sendMessage(message);
    }
}

File: multifiledownloader/src/main/java/com/github/why168/multifiledownloader/call/AsyncConnectCall.kt
package com.github.why168.multifiledownloader.call

import com.github.why168.multifiledownloader.db.DataBaseUtil.updateDownLoadById
import android.annotation.SuppressLint
import android.content.Context
import android.os.Handler
import android.util.Log
import com.github.why168.multifiledownloader.Constants
import com.github.why168.multifiledownloader.DownLoadBean
import com.github.why168.multifiledownloader.DownLoadState
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ExecutorService
import java.util.concurrent.atomic.AtomicBoolean

/**
 * 连接的线程
 *
 * @author Edwin.Wu
 * @version 2017/1/16 14:15
 * @since JDK1.8
 */
@SuppressLint("SimpleDateFormat")
class AsyncConnectCall constructor(
    private val context: Context, private val handler: Handler,
    connectionTaskMap: ConcurrentHashMap<String, AsyncConnectCall>,
    private val downTaskMap: ConcurrentHashMap<String, AsyncDownCall>,
    private val executorService: ExecutorService,
    private val bean: DownLoadBean
) : NickRunnable(
    format = "AndroidHttp %s",
    args = SimpleDateFormat("yyyy-MM-dd-HH-mm-ss-SSS").format(Calendar.getInstance().time)
) {
    private val connectionTaskMap: ConcurrentHashMap<String, AsyncConnectCall>
    private val isRunning: AtomicBoolean = AtomicBoolean(true)

    init {
        this.connectionTaskMap = connectionTaskMap
        this.connectionTaskMap[bean.id] = this
    }

    override fun execute() {
        bean.downloadState = DownLoadState.STATE_CONNECTION.index
        updateDownLoadById(context, bean)
        notifyDownloadStateChanged(bean, DownLoadState.STATE_CONNECTION.index)
        var connection: HttpURLConnection? = null
        try {
            connection = URL(bean.url).openConnection() as HttpURLConnection
            connection.requestMethod = "GET"
            connection.connectTimeout = Constants.CONNECT_TIME
            connection.readTimeout = Constants.READ_TIME
            val responseCode = connection.responseCode
            val contentLength = connection.contentLength
            if (responseCode == HttpURLConnection.HTTP_OK) {
                val ranges = connection.getHeaderField("Accept-Ranges")
                if ("bytes".equals(ranges, ignoreCase = true)) {
                    bean.isSupportRange = true
                }
                bean.totalSize = (contentLength.toString() + "").toLong()
            } else {
                bean.downloadState = DownLoadState.STATE_ERROR.index
            }

//            UpdateDownLoadById(context, bean);
//            notifyDownloadStateChanged(bean, DownLoadState.STATE_CONNECTION.index);
            Log.d("Edwin", "连接成功--isSupportRange = " + bean.isSupportRange)

            // 开始下载咯
            val downLoadTask = AsyncDownCall(context, handler, bean)
            if (downTaskMap[bean.id] == null) {
                downTaskMap[bean.id] = downLoadTask
                executorService.execute(downLoadTask)

                // 移除
                connectionTaskMap.remove(bean.id)
            }
        } catch (e: IOException) {
            e.printStackTrace()
            isRunning.set(false)
            bean.downloadState = DownLoadState.STATE_ERROR.index
            updateDownLoadById(context, bean)
            notifyDownloadStateChanged(bean, DownLoadState.STATE_ERROR.index)
            Log.d("Edwin", "连接失败")
        } finally {
            connection?.disconnect()
        }
    }

    fun cancel() {
        isRunning.set(true)
    }

    fun isCancel(): Boolean {
        return isRunning.get()
    }

    /**
     * 当下载状态发送改变的时候调用
     */
    private fun notifyDownloadStateChanged(bean: DownLoadBean, state: Int) {
        val message = handler.obtainMessage()
        message.obj = bean
        message.what = state
        handler.sendMessage(message)
    }
}